src/ddsrt/src/sync/freertos/sync.c

```diff
@@ -0,0 +1,505 @@
+ * v. 1.0 which is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR BSD-3-Clause
+ */
+#include <FreeRTOS.h>
+#include <task.h>
+#include <assert.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "dds/ddsrt/heap.h"
+#include "dds/ddsrt/log.h"
+#include "dds/ddsrt/sync.h"
+#include "dds/ddsrt/timeconv.h"
+
+void ddsrt_mutex_init(ddsrt_mutex_t *mutex)
+{
+  SemaphoreHandle_t sem;
+
+  assert(mutex != NULL);
+
+  if ((sem = xSemaphoreCreateMutex()) == NULL) {
+    abort();
+  }
+
+  (void)memset(mutex, 0, sizeof(*mutex));
+  mutex->sem = sem;
+}
+
+void ddsrt_mutex_destroy(ddsrt_mutex_t *mutex)
+{
+  assert(mutex != NULL);
+
+  vSemaphoreDelete(mutex->sem);
+  (void)memset(mutex, 0, sizeof(*mutex));
+}
+
+static bool
+mutex_lock(ddsrt_mutex_t *mutex, int blk)
+{
+  assert(mutex != NULL);
+
+  if (xSemaphoreTake(mutex->sem, (blk == 1 ? portMAX_DELAY : 0)) != pdPASS) {
+    DDS_TRACE("Failed to lock 0x%p", mutex);
+    /* xSemaphoreTake will only return pdFAIL on timeout. The wait will be
+       indefinite if INCLUDE_vTaskSuspend is set to 1 in FreeRTOSConfig.h
+       and portMAX_DELAY was passed. */
+    assert(blk == 0);
+    return false;
+  }
+
+  return true;
+}
+
+void ddsrt_mutex_lock(ddsrt_mutex_t *mutex)
+{
+  if (!mutex_lock(mutex, 1)) {
+    abort();
+  }
+}
+
+bool ddsrt_mutex_trylock(ddsrt_mutex_t *mutex)
+{
+  return mutex_lock(mutex, 0);
+}
+
+void ddsrt_mutex_unlock(ddsrt_mutex_t *mutex)
+{
+  assert(mutex != NULL);
+
+  if (xSemaphoreGive(mutex->sem) != pdPASS) {
+    DDS_TRACE("Failed to unlock 0x%p", mutex->sem);
+    abort();
+  }
+}
+
+static dds_retcode_t
+cond_resize(ddsrt_cond_t *cond, size_t len)
+{
+  dds_retcode_t rc = 0;
+  size_t off;
+  TaskHandle_t *buf = NULL;
+  static const size_t sz = sizeof(*cond->tasks.buf);
+
+  assert(cond != NULL);
+
+  /* Free buffer if required number of slots is zero. */
+  if (len == 0) {
+    ddsrt_free(cond->tasks.buf);
+    (void)memset(&cond->tasks, 0, sizeof(cond->tasks));
+  } else {
+    if (len > cond->tasks.len) {
+      off = cond->tasks.len;
+    } else {
+      off = len;
+    }
+
+    /* Resize buffer if required number of slots is more than available or
+       amount of space freed up is worth the effort. */
+    if (len > cond->tasks.len || len < (cond->tasks.len - 1)) {
+      buf = ddsrt_realloc(cond->tasks.buf, len * sz);
+      if (buf == NULL) {
+        rc = DDS_RETCODE_OUT_OF_RESOURCES;
+      } else {
+        cond->tasks.buf = buf;
+        cond->tasks.len = len;
+      }
+    }
+
+    /* Initialize memory that should not be used anymore/yet to zero. */
+    if (off < cond->tasks.len) {
+      (void)memset(cond->tasks.buf + off, 0, (cond->tasks.len - off) * sz);
+    }
+  }
+
+  return rc;
+}
+
+static dds_retcode_t
+cond_push(ddsrt_cond_t *cond, TaskHandle_t task)
+{
+  size_t cnt;
+
+  assert(cond != NULL);
+  assert(task != NULL);
+
+  /* Ensure the task is not listed. */
```

Comment:
That's a somewhat misleading comment: you're searching for it, but then sort-of allows for the case where it is listed